### PR TITLE
Add TCP+TLS Healthchecks

### PR DESCRIPTION
**Pull request by pgporada**
### Description

Allows TCP healthchecks to optionally connect to a listening TCP server over TLS.

Fixes https://github.com/hashicorp/consul/issues/3936.

### PR Checklist

* [X] updated test coverage
* [X] external facing docs updated
* [ ] appropriate backport labels added
* [ ] not a security concern - well, sort of?

### CODE REVIEW

File: tlsutil/config.go
Comment (beautifulentropy) [inline tlsutil/config.go:881]:
```suggestion
```

File: website/content/docs/ecs/configuration-reference.mdx
Comment (beautifulentropy) [inline website/content/docs/ecs/configuration-reference.mdx:195]:
```suggestion
| `tlsSkipVerify` | `boolean` | optional | Determines if the check verifies the chain and hostname of the certificate presented by the server being checked. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. Supported check types: `HTTP`, `H2Ping`, `gRPC`, and `TCP`|
```

**Review by beautifulentropy — CHANGES_REQUESTED** [review]
Looks great, just some small nits and some additional documentation that needs updating.


Update the the documentation, here: `website/content/docs/services/configuration/checks-configuration-reference.mdx`

Modify line `39`:
```markdown
| `tls_skip_verify` | Boolean value that determines if the check verifies the chain and hostname of the certificate that the server presents. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. | <li>HTTP </li> <li>H2Ping </li> <li>gRPC </li> <li>TCP </li> |
```

Add line below `46` (under `tcp:`):
```markdown
| `tcp_use_tls` | Boolean value that enables TLS for TCP checks when set to `true`. | <li>gRPC </li> |
```

**Comment by pgporada** [comment]
Thanks for the review, all the doc changes have been completed.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:50]: Revert this line change.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:43]: Revert this line change.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:35]: Revert this line change.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:33]: Revert this line change.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:19]: Revert both of these line changes.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:16]: Revert this line change.

File: website/content/docs/ecs/configuration-reference.mdx
Comment (beautifulentropy) [inline website/content/docs/ecs/configuration-reference.mdx:195]: My bad, use verbiage that's more consistent with the other lines in this file: 

```suggestion
| `tlsSkipVerify` | `boolean` | optional | Specifies if the check should verify the chain and hostname of the certificate presented by the server being checked. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. Supported check types: `HTTP`, `H2Ping`, `gRPC`, and `TCP`|
```

Comment (beautifulentropy) [inline website/content/docs/ecs/configuration-reference.mdx:195]: My bad, use verbiage that's more consistent with the other lines in this file: 

```suggestion
| `tlsSkipVerify` | `boolean` | optional | Specifies if the check should verify the chain and hostname of the certificate presented by the server being checked. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. Supported check types: `HTTP`, `H2Ping`, `gRPC`, and `TCP`|
```

**Review by beautifulentropy — CHANGES_REQUESTED** [review]

File: agent/agent_test.go
Comment (beautifulentropy) [inline agent/agent_test.go:991]: Here and elsewhere, we should use `require.NoError()`.

```suggestion
	require.NoError(t, err)
```

File: agent/agent_test.go
Comment (beautifulentropy) [inline agent/agent_test.go:1009]: Here and elsewhere, we should drop these log lines.

```suggestion
```

File: agent/agent_test.go
Comment (beautifulentropy) [inline agent/agent_test.go:1101]: The implementation you have here is great for debugging your implementation. However, for the final test case, we should go with something smaller that uses more of the standard library:

```suggestion
func startMockTLSServer(t *testing.T) (addr string, closeFunc func()) {
	// Load certificates
	cert, err := tls.LoadX509KeyPair("../test/key/ourdomain_server.cer", "../test/key/ourdomain_server.key")
	require.NoError(t, err)

	// Create a certificate pool
	rootCertPool := x509.NewCertPool()
	caCert, err := os.ReadFile("../test/ca/root.cer")
	require.NoError(t, err)
	rootCertPool.AppendCertsFromPEM(caCert)

	// Configure TLS
	config := &tls.Config{
		Certificates: []tls.Certificate{cert},
		ClientAuth:   tls.RequireAndVerifyClientCert,
		ClientCAs:    rootCertPool,
	}

	// Start TLS server
	ln, err := tls.Listen("tcp", "127.0.0.1:0", config)
	require.NoError(t, err)

	go func() {
		for {
			conn, err := ln.Accept()
			if err != nil {
				return
			}
			go func(c net.Conn) {
				io.Copy(c, c)
				c.Close()
			}(conn)
		}
	}()

	return ln.Addr().String(), func() { ln.Close() }
}

func TestAgent_AddServiceWithTCPTLSCheck(t *testing.T) {
	t.Parallel()
	dataDir := testutil.TempDir(t, "agent")
	a := NewTestAgent(t, `
		data_dir = "`+dataDir+`"
		enable_agent_tls_for_checks = true
		datacenter = "dc1"
		tls {
			defaults {
				ca_file   = "../test/ca/root.cer"
				cert_file = "../test/key/ourdomain_server.cer"
				key_file  = "../test/key/ourdomain_server.key"
			}
		}
	`)
	defer a.Shutdown()
	testrpc.WaitForTestAgent(t, a.RPC, "dc1")

	// Start mock TCP+TLS server
	addr, closeServer := startMockTLSServer(t)
	defer closeServer()

	check := &structs.HealthCheck{
		Node:    "foo",
		CheckID: "arbitraryTCPServerTLSCheck",
		Name:    "arbitraryTCPServerTLSCheck",
		Status:  api.HealthCritical,
	}

	chkType := &structs.CheckType{
		TCP:           addr,
		TCPUseTLS:     true,
		TLSServerName: "server.dc1.consul",
		Interval:      5 * time.Second,
	}

	err := a.AddCheck(check, chkType, false, "", ConfigSourceLocal)
	require.NoError(t, err)

	// Retry until the healthcheck is passing.
	retry.Run(t, func(r *retry.R) {
		status := getCheck(a, "arbitraryTCPServerTLSCheck")
		if status.Status != api.HealthPassing {
			r.Fatalf("bad: %v", status.Status)
		}
	})
}

```

**Review by beautifulentropy — CHANGES_REQUESTED** [review]

**Comment by trujillo-adam** [comment]
I don't see any updates to the main configuration reference page for health checks: https://developer.hashicorp.com/consul/docs/services/configuration/checks-configuration-reference

Are these updates not relevant?

**Comment by pgporada** [comment]
That page is relevant, it was forgotten in the mix. Apologies.

**Comment by pgporada** [comment]
@trujillo-adam Turns out that the file that builds the docs at the link you sent was already in my PR. See https://github.com/hashicorp/consul/pull/18381/files#diff-c4d30ce7367e53ab22681c3938502abd8fefd01546c6c85a4454c21b45bd7100R47.

File: agent/checks/check.go
Comment (loshz) [inline agent/checks/check.go:697]: I think we can simplify this whole function if we use an interface for the conn:
```go
func (c *CheckTCP) check() {
	var conn io.Closer
	var err error

	if c.TLSClientConfig == nil {
		conn, err = c.dialer.Dial(`tcp`, c.TCP)
	} else {
		conn, err = tls.DialWithDialer(c.dialer, `tcp`, c.TCP, c.TLSClientConfig)
	}

	if err != nil {
		c.Logger.Warn("Check TCP connection failed",
			"check", c.CheckID.String(),
			"error", err,
		)
		c.StatusHandler.updateCheck(c.CheckID, api.HealthCritical, err.Error())
		return
	}

	conn.Close()
	c.StatusHandler.updateCheck(c.CheckID, api.HealthPassing, fmt.Sprintf("TCP connect %s: Success", c.TCP))
}
```

WDYT?

File: website/content/docs/ecs/configuration-reference.mdx
Comment (loshz) [inline website/content/docs/ecs/configuration-reference.mdx:195]: I am ok with this new wording, but I'll will have to defer to someone from our security posture to see what they think.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx
Comment (loshz) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:42]: Revert this change.

**Review by loshz — CHANGES_REQUESTED** [review]
Hi @pgporada - thanks for working on this, and sorry it took so long for me to review! 

This is a very useful feature that I know has been requested for a while, so I appreciate you taking the time to work on it.

I've left a couple of comments, and need to also get another eng from Hashi to validate a few things, but I can do that.

Feel free to ping me if you have any questions!

File: agent/checks/check.go
Comment (pgporada) [reply agent/checks/check.go:697]: That's a great idea, I hadn't considered that was possible.

**Comment by pgporada** [comment]
Thank you for taking the time to review this. :)

File: agent/agent_test.go
Comment (pgporada) [inline agent/agent_test.go:983]: For lines 982 and 983, I chose not to generate new key material. The `tls.RequireAndVerifyClientCert` is working by virtue of the Consul Agent sending a TLS cert with a SAN that matches a SAN loaded by the mockTLSServer. If you want me to generate new key material and change this, that's fine by me.

File: agent/agent_test.go
Comment (loshz) [inline agent/agent_test.go:25]: Looks like a merge conflict was missed here?

Comment (pgporada) [reply agent/agent_test.go]: Fixed, sorry about that.

**Comment by loshz** [comment]
@david-yu do we want to backport this? Or is it deemed a "new feature"?

**Comment by david-yu** [comment]
I would backport to just 1.16.x just to limit blast radius if there are issues. I think it would be good to backport to at least get usage of this feature in our latest stable branch.

**Comment by loshz** [comment]
@pgporada thanks for fixing the merge conflict and addressing the pr feedback. One last thing we need to do is add a changelog file (`.changelog/18381.txt`), similar to something like [this](https://github.com/hashicorp/consul/blob/main/.changelog/18617.txt), that outlines the specific change made here.

**Comment by pgporada** [comment]
@loshz Done, thanks for the heads up on that.

**Review by loshz — APPROVED** [review]
LGTM - I appreciate you taking the time to see this through and making the appropriate changes!

FYI, once this gets merged, a follow up PR will be generated in order to backport this to 1.16 versions. It should get auto merged, but I'll keep an eye on it in case it doesn't.

**Pull request merged**
